File: packages/python/elizaos/features/basic_capabilities/providers/providers_list.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.prompt_compression import get_prompt_provider_description
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_providers_list(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    selection_hints = [
        "images, attachments, or visual content -> ATTACHMENTS",
        "specific people or agents -> ENTITIES",
        "connections between people -> RELATIONSHIPS",
        "factual lookup -> FACTS",
        "world or environment context -> WORLD",
    ]
    provider_info: list[dict[str, str | bool]] = []

    for provider in runtime.providers:
        provider_info.append(
            {
                "name": provider.name,
                "description": get_prompt_provider_description(provider, runtime),
                "dynamic": getattr(provider, "dynamic", True),
            }
        )

    if not provider_info:
        return ProviderResult(
            text="# Available Providers\nproviders[0]:\n- none",
            values={"providerCount": 0},
            data={"providers": []},
        )

    formatted_providers = "\n".join(f"- {p['name']}: {p['description']}" for p in provider_info)
    formatted_hints = "\n".join(f"- {hint}" for hint in selection_hints)
    text = (
        f"# Available Providers\nproviders[{len(provider_info)}]:\n{formatted_providers}\n"
        f"provider_hints[{len(selection_hints)}]:\n{formatted_hints}"
    )

    return ProviderResult(
        text=text,
        values={
            "providerCount": len(provider_info),
            "providerNames": [p["name"] for p in provider_info],
        },
        data={
            "providers": provider_info,
        },
    )


providers_list_provider = Provider(
    name="PROVIDERS",
    description="Available context providers",
    get=get_providers_list,
    dynamic=False,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/providers_list.rs
//! PROVIDERS provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("PROVIDERS"));

/// Provider for listing available providers.
pub struct ProvidersListProvider;

#[async_trait]
impl Provider for ProvidersListProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(false)
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let selection_hints = [
            "images, attachments, or visual content -> ATTACHMENTS",
            "specific people or agents -> ENTITIES",
            "connections between people -> RELATIONSHIPS",
            "factual lookup -> FACTS",
            "world or environment context -> WORLD",
        ];
        // Get providers from the basic_capabilities plugin itself
        let providers = super::all_providers();

        if providers.is_empty() {
            return Ok(
                ProviderResult::new("# Available Providers\nproviders[0]:\n- none")
                    .with_value("providerCount", 0i64),
            );
        }

        let provider_info: Vec<serde_json::Value> = providers
            .iter()
            .map(|p| {
                serde_json::json!({
                    "name": p.name(),
                    "description": p.description(),
                    "dynamic": p.is_dynamic()
                })
            })
            .collect();

        let formatted: Vec<String> = providers
            .iter()
            .map(|p| format!("- {}: {}", p.name(), p.description()))
            .collect();

        let text = format!(
            "# Available Providers\nproviders[{}]:\n{}\nprovider_hints[{}]:\n{}",
            providers.len(),
            formatted.join("\n"),
            selection_hints.len(),
            selection_hints
                .iter()
                .map(|hint| format!("- {}", hint))
                .collect::<Vec<String>>()
                .join("\n")
        );

        let names: Vec<&str> = providers.iter().map(|p| p.name()).collect();

        Ok(ProviderResult::new(text)
            .with_value("providerCount", providers.len() as i64)
            .with_data(
                "providerNames",
                serde_json::to_value(&names).unwrap_or_default(),
            )
            .with_data(
                "providers",
                serde_json::to_value(&provider_info).unwrap_or_default(),
            ))
    }
}
